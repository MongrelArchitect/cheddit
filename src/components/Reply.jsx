import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { doc, getDoc, updateDoc } from 'firebase/firestore';
import {
  getDownloadURL, getStorage, ref, uploadBytes,
} from 'firebase/storage';
import database from '../util/firestore';

async function loadBoard(id) {
  const boardRef = doc(database, 'boards', id);
  const boardSnap = await getDoc(boardRef);
  return boardSnap.data();
}

async function getNewPostNumber(id) {
  const board = await loadBoard(id);
  const lastPost = Object.keys(board.posts);
  return +lastPost[lastPost.length - 1] + 1;
}

export default function Reply({ board, readDatabase, thread }) {
  const [enabled, setEnabled] = useState(false);
  const [postAuthor, setPostAuthor] = useState('');
  const [postContent, setPostContent] = useState('');
  const [file, setFile] = useState(null);

  const changeAuthor = (event) => {
    setPostAuthor(event.target.value);
  };

  const changeContent = (event) => {
    setPostContent(event.target.value);
  };

  const changeFile = (event) => {
    setFile(event.target.files[0]);
  };

  const enableForm = () => {
    setEnabled(!enabled);
  };

  const uploadImage = async (newPostNumber) => {
    const storage = getStorage();
    const extension = file.name.match(/\.[a-zA-Z0-9]+$/).join();
    const imageRef = ref(storage, `${board}/${newPostNumber}${extension}`);
    let downloadURL = null;
    try {
      const snapshot = await uploadBytes(imageRef, file);
      downloadURL = await getDownloadURL(snapshot.ref);
    } catch (error) {
      console.error(`Error uploading file: ${error}`);
    }
    return downloadURL;
  };

  const submitPost = async () => {
    const newPost = {
      author: postAuthor === '' ? null : postAuthor,
      content: postContent,
      image: null,
      replies: [],
      subject: null,
      thread,
      time: Date.now(),
    };
    const newPostNumber = await getNewPostNumber(board);
    const update = {};
    const updateKey = `posts.${newPostNumber}`;
    update[updateKey] = newPost;

    if (file) {
      const URL = await uploadImage(newPostNumber);
      update[updateKey].image = URL;
    }

    const boardRef = doc(database, 'boards', board);
    await updateDoc(boardRef, update);
    setPostAuthor('');
    setPostContent('');
    setFile(null);
    enableForm();
    await readDatabase();
  };

  if (enabled) {
    return (
      <div className="reply-container">
        <form aria-label="reply form" className="reply-form">
          <label htmlFor="post-author">
            Name:
            <input
              id="post-author"
              name="post-author"
              onChange={changeAuthor}
              type="text"
              placeholder="Anonymous"
              value={postAuthor}
            />
          </label>
          <label htmlFor="post-content">
            Comment:
            <textarea
              id="post-content"
              name="post-content"
              onChange={changeContent}
              rows="6"
              value={postContent}
            />
          </label>
          <label htmlFor="post-image">
            Image:
            <input
              accept="image/jpeg, image/gif, image/png"
              id="post-image"
              name="post=image"
              onChange={changeFile}
              type="file"
            />
          </label>
          <div className="buttons">
            <button onClick={enableForm} type="button">
              CANCEL
            </button>
            <button onClick={submitPost} type="button">
              POST
            </button>
          </div>
        </form>
        <Link to={`/${board}`}>{`Back to /${board}/`}</Link>
      </div>
    );
  }

  return (
    <div className="reply-container">
      <button onClick={enableForm} type="button">
        [ Post a Reply ]
      </button>
      <Link to={`/${board}`}>{`Back to /${board}/`}</Link>
    </div>
  );
}
