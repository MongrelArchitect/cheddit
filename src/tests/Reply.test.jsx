import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Reply from '../components/Reply';

describe('Reply component', () => {
  it('displays reply button', () => {
    render(
      <BrowserRouter>
        <Reply />
      </BrowserRouter>,
    );
    expect(
      screen.getByRole('button', { name: 'Post a Reply' }),
    ).toBeTruthy();
  });

  it('displays reply form when button clicked', () => {
    render(
      <BrowserRouter>
        <Reply />
      </BrowserRouter>,
    );
    act(() => {
      userEvent.click(screen.getByRole('button', { name: 'Post a Reply' }));
    });
    expect(screen.getByRole('form', { name: 'reply form' })).toBeTruthy();
  });

  it('displays correct input fields', () => {
    render(
      <BrowserRouter>
        <Reply />
      </BrowserRouter>,
    );
    act(() => {
      userEvent.click(screen.getByRole('button', { name: 'Post a Reply' }));
    });
    expect(screen.getByLabelText('Name:')).toBeTruthy();
    expect(screen.getByLabelText('Comment:')).toBeTruthy();
    expect(screen.getByLabelText('Image:')).toBeTruthy();
  });
});
