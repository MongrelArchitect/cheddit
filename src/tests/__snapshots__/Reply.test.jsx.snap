// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reply component displays reply button 1`] = `
<div>
  <div
    class="reply-container"
  >
    <form
      aria-label="reply form"
      class="reply-form"
    >
      <label
        for="post-author"
      >
        Name:
        <span
          id="post-author"
        >
          Anonymous
        </span>
      </label>
      <label
        for="post-content"
      >
        Comment:
        <textarea
          id="post-content"
          name="post-content"
          rows="6"
        />
      </label>
      <label
        for="post-image"
      >
        Image:
        <input
          accept="image/jpeg, image/gif, image/png"
          id="post-image"
          name="post=image"
          type="file"
        />
      </label>
      <span
        class="error"
        hidden=""
      />
      <div
        class="buttons visible"
      >
        <button
          type="button"
        >
          CANCEL
        </button>
        <button
          type="button"
        >
          POST
        </button>
      </div>
      <div
        class="loading hidden"
      >
        <span>
          Uploading...
        </span>
        <img
          alt=""
          src="loading.gif"
        />
      </div>
    </form>
    <a
      href="/undefined"
    >
      Back to /undefined/
    </a>
  </div>
</div>
`;
